alf/environments/carla_sensors.py

```diff
@@ -705,6 +874,164 @@ def transform_to_geolocation(self, location: carla.Location):
+
+        logging.info(
+            "Found %d traffic lights" % len(self._traffic_light_actors))
+
+        self._traffic_light_centers = np.array(traffic_light_centers,
+                                               np.float32)
+        np_traffic_light_waypoints = []
+        for waypoints in traffic_light_waypoints:
+            pad = max_num_traffic_light_waypoints - len(waypoints)
+            if pad > 0:
+                waypoints.extend([dummy_waypoint] * pad)
+            np_traffic_light_waypoints.append(
+                alf.nest.map_structure(lambda *x: np.stack(x), *waypoints))
+
+        self._traffic_light_waypoints = alf.nest.map_structure(
+            lambda *x: np.stack(x), *np_traffic_light_waypoints)
+
+        self._traffic_light_centers = np.array(traffic_light_centers)
+
+    def on_tick(self):
+        """Should be called after every world tick() to update data."""
+        self._traffic_light_states = np.array(
+            [a.state for a in self._traffic_light_actors], dtype=np.int)
+
+    def _get_traffic_light_waypoints(self, traffic_light):
+        # Copied from RunningRedLightTest.get_traffic_light_waypoints() in
+        # https://github.com/carla-simulator/scenario_runner/blob/master/srunner/scenariomanager/scenarioatomics/atomic_criteria.py
+        base_transform = traffic_light.get_transform()
+        base_rot = base_transform.rotation.yaw
+        area_loc = base_transform.transform(
+            traffic_light.trigger_volume.location)
+
+        # Discretize the trigger box into points
+        area_ext = traffic_light.trigger_volume.extent
+        x_values = np.arange(-0.9 * area_ext.x, 0.9 * area_ext.x,
+                             1.0)  # 0.9 to avoid crossing to adjacent lanes
+
+        area = []
+        for x in x_values:
+            point = _rotate_point(carla.Vector3D(x, 0, area_ext.z), base_rot)
+            point_location = area_loc + carla.Location(x=point.x, y=point.y)
+            area.append(point_location)
+
+        # Get the waypoints of these points, removing duplicates
+        ini_wps = []
+        for pt in area:
+            wpx = self._map.get_waypoint(pt)
+            # As x_values are arranged in order, only the last one has to be checked
+            if not ini_wps or ini_wps[-1].road_id != wpx.road_id or ini_wps[
+                    -1].lane_id != wpx.lane_id:
+                ini_wps.append(wpx)
+
+        # Advance them until the intersection
+        wps = []
+        for wpx in ini_wps:
+            while not wpx.is_intersection:
+                next_wp = wpx.next(0.5)[0]
+                if next_wp and not next_wp.is_intersection:
+                    wpx = next_wp
+                else:
+                    break
+            wps.append(wpx)
+
+        # self._draw_waypoints(wps, vertical_shift=1.0, persistency=50000.0)
+
+        return area_loc, wps
+
+    def is_running_red_light(self, actor):
+        """Whether actor is running red light.
```

Comment:
docstring is broken

Reply:
Oh I meant the line below has incorrect format.

Reply:
Did you mean that the URL is broken into two lines? That's just because github has to wrap the long line.

Reply:
never mind, I was talking about the code and url format.